pkg/arvo/neo/cod/std/src/fil/messenger.md

```diff
@@ -0,0 +1,649 @@
+    :~  :-  (snoc here.bowl %pub)
+        [%make %message-pub ~ ~]
+        ::
+        :-  (snoc here.bowl %sub)
+        [%make %message-sub ~ (malt ~[[%pub (snoc here.bowl %pub)]])]
+    ==
+  ::  otherwise, I've been created as an invitee to
+  ::  someone else's groupchat
+  ?>  =(%groupchat-diff p.u.old)
+  =/  poke  !<(groupchat-diff q.u.old)
+  ?+    -.poke  !!
+      %invited
+    :_  :-  %groupchat
+        !>([~ ~ host.poke])
+    :~  :-  (snoc here.bowl %sub)
+        [%make %message-sub ~ (malt ~[[%pub (snoc host.poke %pub)]])]
+        ::
+        :-  host.poke
+        [%poke groupchat-diff/!>([%acked ~])]
+    ==
+  ==
+```
+
+#### +poke
+In the `+poke` arm we handle `%invite`, `%remove`, `%acked`, and `%post-to-host`, which is mostly a wrapper around `%host-to-pub`.
+
+Even though it doesn't handle messages — just access control — `/imp/groupchat` has some state to manage. Moreso than antyhing we've seen before, the below should look a lot like a Gall agent.
+
+{what's the heuristic for "should I store this data in "top-level" shrub state or in the namespace below it?}
+
+```hoon
+++  poke
+  |=  [=stud:neo vax=vase]
+  ^-  (quip card:neo pail:neo)
+  ?>  =(%groupchat-diff stud)
+  =/  sta  !<(groupchat q.state)
+  =/  poke  !<(groupchat-diff vax)
+  ?+    -.poke  !!
+      :: if I'm the host, poke someone's provider to invite them to chat
+      %invite
+    ?>  =(our ship.src):bowl
+    ?<  (~(has in members.sta) ship.poke)
+    ::  ?>  =(our.bowl ->.host.sta) :: XX need @p, have @t ?
+    :_  :-  %groupchat
+        !>(sta(pending (~(put in pending.sta) ship.poke)))
+    :~  :-  provider.poke
+        [%poke groupchat-diff/!>([%invited here.bowl])]
+    ==
+  ::
+      ::  remove someone from chat. this only removes their ability to post;
+      ::  they'll still be receiving new messages!
+      %remove
+    ?>  =(our ship.src):bowl
+    ?>  (~(has in members.sta) ship.poke)
+    :-  ~
+    :-  %groupchat
+    !>  %=  sta
+          pending  (~(del in pending.sta) ship.src.bowl)
+          members  (~(del in members.sta) ship.src.bowl)
+        ==
+  ::
+      :: when invitee acks, remove them from pending
+      :: and add them to pub's permissions
+      %acked
+    ?>  (~(has in pending.sta) ship.src.bowl)
+    :-  ~
+    :-  %groupchat
+    !>  %=  sta
+          pending  (~(del in pending.sta) ship.src.bowl)
+          members  (~(put in members.sta) ship.src.bowl)
+        ==
+  ::
+      %post-to-host
+    :_  state
+    :~  :-  host.sta
+        [%poke groupchat-diff/!>([%host-to-pub text.poke])]
+    ==
+  ::
+      %host-to-pub
+    ?>  (~(has in members.sta) ship.src.bowl)
+    :_  state
+    :~  :-  (snoc here.bowl %pub)
+        [%poke message/!>([ship.src.bowl now.bowl text.poke])]
+    ==
+  ==
+```
+
+### Messenger shrub
+`/imp/messenger` is the top-level interface through which users can create, post in, and manage groupchats and DMs. This is the only part of the system that has a UI.
```

Comment:
"This is the only part of the system that has a UI." not quite true, there are also con/dm.hoon and con/groupchat.hoon